src/transformers/modeling_utils.py

```diff
@@ -3405,6 +3405,17 @@ def save_pretrained(
         # (initially introduced with TimmWrapperModel to remove prefix and make checkpoints compatible with timm)
         state_dict = self._fix_state_dict_keys_on_save(state_dict)
 
+        # if using tensor parallel we need to gather the tensors in state dict
+        gathered_state_dict = {}
+        for key, value in state_dict.items():
+            if hasattr(value, "_local_tensor"):
+                gathered_state_dict[key] = value.to_local().cpu()
```

Comment:
memory will explode no? this should happen in the function that write the files to make sure you save bits by bits